### PR TITLE
bug/#1280 - possibility to draw Polylines using Path

### PR DESCRIPTION
**Pull request by monsieurtanuki**
New classes:
* `PolyOverlayWithIW`: extending `OverlayWithIW`, extended by both `Polyline` and `Polygon` given the number of common methods
* `SampleDrawPolylineAsPath`: sample similar to `SampleDrawPolyline`, but where `Polyline`s are drawing in the "path" mode (and not in the "lines" mode)

Impacted classes:
* `CustomPaintingSurface`: created mode `PolylineAsPath`; put alpha in the color in order to see the vertices for `Polyline`s
* `LinearRing`: added the `boolean` member `mClosed` in order to make both `Polyline` and `Polygon` draw using a `Path`; new constructor for `Path` with a `boolean pClosed` parameter; taking into account the `mClosed` member when building the `Path`
* `SampleFactory`: added `SampleDrawPolylineAsPath` just after `SampleDrawPolyline`
* `Polygon`: now extends new class `PolyOverlayWithIW`
* `Polyline`: now extends new class `PolyOverlayWithIW`

Impacted classes - slightly unrelated refactoring regarding `Paint`s:
* `Bug382Crash`
* `HeatMap`
* `LatLonGridlineOverlay`
* `LatLonGridlineOverlay2`
* `MarkerDrag`
* `MilStdMultipointOverlay`
* `OsmMapShapeConverter`
* `SampleMilestonesNonRepetitive`
* `SampleOsmPath`
* `SampleRace`
* `SampleZoomToBounding`

### CODE REVIEW

**Comment by conej730** [comment]
Thanks @monsieurtanuki this is great.

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/LinearRing.java [hunk 1]
```diff
@@ -194,9 +208,11 @@ PointL buildPathPortion(final Projection pProjection,
 			getBestOffset(pProjection, offset);
 		}
 		mSegmentClipper.init();
-		clipAndStore(pProjection, offset, true, pStorePoints, mSegmentClipper);
+		clipAndStore(pProjection, offset, mClosed, pStorePoints, mSegmentClipper);
 		mSegmentClipper.end();
-		mPath.close();
+		if (mClosed) {
+		    mPath.close();
+        }
```
Comment (conej730) [inline osmdroid-android/src/main/java/org/osmdroid/views/overlay/LinearRing.java]: tab issue

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		}
+		if (pUsePath) {
+			mPath = new Path();
+			mLineDrawer = null;
+			mOutline = new LinearRing(mPath, pClosePath);
+		} else {
+			mPath = null;
+			mLineDrawer = new LineDrawer(256);
+			mOutline = new LinearRing(mLineDrawer);
+			////mOutline.clearPath();
```
Comment (conej730) [inline osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java]: I think this comment could be removed?

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+			milestoneManager.init();
+			milestoneManager.setDistances(mOutline.getDistances());
+			for (final PointL point : mOutline.getPointsForMilestones()) {
+				milestoneManager.add(point.x, point.y);
+			}
+			milestoneManager.end();
+		}
+
+		for (final MilestoneManager milestoneManager : mMilestoneManagers) {
+			milestoneManager.draw(pCanvas);
```
Comment (conej730) [inline osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java]: Can the draw be done in the previous loop after the milestone manager is populated, or does it cause delayed draw behavior that way?

**Comment by spyhunter99** [comment]
let's hold off on merging this until the rest of the bugs are ironed out.

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java (hunk 3 above)
Comment (monsieurtanuki) [reply osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java]: We need these two steps.
The first step computes the actual "milestones", which can be something like a milestone every km, or 5 equidistant milestones for the whole length of the polyline. Therefore we need to parse everything first. Then draw, in the second step.

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java (hunk 2 above)
Comment (monsieurtanuki) [reply osmdroid-android/src/main/java/org/osmdroid/views/overlay/PolyOverlayWithIW.java]: Fair enough, `mOutline.clearPath();` should be executed and not commented.

File: osmdroid-android/src/main/java/org/osmdroid/views/overlay/LinearRing.java (hunk 1 above)
Comment (monsieurtanuki) [reply osmdroid-android/src/main/java/org/osmdroid/views/overlay/LinearRing.java]: OK

**Comment by conej730** [comment]
We've been testing with this branch for a bit now.  So far no issues have arisen.

**Comment by Ruediger34** [comment]
I am totally fine with this implementation. 
Tested with no performance related impacts or other issues in my use case. 

Would be nice to this in the next release. 

Thank you @monsieurtanuki

**Comment by spyhunter99** [comment]
roger that. i want to cut another version before merging this, but it will be first on the list

**Comment by monsieurtanuki** [comment]
@spyhunter99 Any progress on a next version?

**Comment by spyhunter99** [comment]
i can cut it today but we still have that one bug with the stack overflow issue

**Comment by monsieurtanuki** [comment]
> i can cut it today but we still have that one bug with the stack overflow issue

Do you mean only #1299?

**Comment by spyhunter99** [comment]
yup

**Comment by spyhunter99** [comment]
merge conflicts

**Comment by monsieurtanuki** [comment]
solved conflicts

**Pull request merged**